(WIP) [762] Update working pattern to allow for more than one

robbiepaul opened the pull request:
## Trello card URL:

https://trello.com/c/4owuXmAO/762-update-working-pattern-options-so-there-is-more-flexibility-around-types-of-role-reflecting-the-needs-of-jobseekers-schools

## Changes in this PR:

- Add a working_patterns array column for array enums
- Rake task to migrate all vacancies to use new enums
- Update hiring staff UI to reflect new column and allow multiple selections
- Update ElasticSearch mappings
- Fixes to some specs

## Screenshots of UI changes:

### Before
<img width="613" alt="Screenshot 2019-03-28 at 19 16 50" src="https://user-images.githubusercontent.com/2804149/55186310-25a09e80-518e-11e9-8ead-00285f2f287c.png">

### After
![Screenshot 2019-04-02 at 15 03 33](https://user-images.githubusercontent.com/2804149/55408728-83e7cb80-5558-11e9-9066-b8f354dbf05f.png)


## Next steps:

- [x] Terraform deployment required?

robbiepaul commented:
@pezholio I opened this PR in favour of the other one as the paradigm has changed. There's a similar amount of changes however there's far less complication. 

I also added another spec for hiring staff publishing jobs with multiple working patterns.

robbiepaul commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Update 'working pattern' options so there is more flexibility around types of role reflecting the needs of jobseekers/schools](https://trello.com/c/4owuXmAO/762-update-working-pattern-options-so-there-is-more-flexibility-around-types-of-role-reflecting-the-needs-of-jobseekers-schools)

erbridge commented on app/presenters/vacancy_presenter.rb:
```diff
@@ -121,11 +121,11 @@ def flexible_working
 
   def working_pattern
-    model.working_pattern.sub('_', ' ').humanize
+    model.working_patterns.join(', ').sub('_', ' ').humanize
```
Doesn't `humanize` already include the underscore substitution?

robbiepaul replied:
Yea, good catch 👍 In fact when looking at it now, I think I need to apply `humanize` to each of the working patterns not both together. I'll add a commit

erbridge commented on lib/tasks/data.rake:
```diff
@@ -26,4 +26,16 @@ namespace :data do
+      Vacancy.where(working_patterns: nil)
+             .where.not(working_pattern: nil)
+             .each do |vacancy|
+        vacancy.working_patterns = [vacancy.working_pattern]
```
Do we need some tests to ensure we're not using the old `working_pattern` anywhere if it's still set?

robbiepaul replied:
I think we have a potential race condition where we need to migrate the existing working pattern to the new column, then remove the old one completely in a separate PR

erbridge replied:
Yeah, I guessed that was why the old field was still there. There will be an intermediary time when we have both fields, however, and during that time, it's plausible that we could be reading or writing to the old field when we mean to hit the new one. I'm not sure where's best to make sure that isn't happening. My instict was to check now, since if we have accidental writes, that would require the data migration to be run again after fixing whatever issue was there, but happy to be overruled :).

robbiepaul replied:
Yeah I think you're right. I've done a search for anywhere where we might be using the old one and can't see anything. We can have some extra confidence from the level of coverage our tests have but maybe it needs something else?

robbiepaul commented:
I've added a commit that "humanizes" the working patterns in the right way and rebased.

I hate to leave a project with PRs unmerged but I hope I've left it in an ok state for the team.

erbridge commented:
After deploy and running the rake task, we need to remove the old `working_pattern` (singular) field.

erbridge commented:
Replaced by #920.
